### PR TITLE
Add tests and functionality for OCSP request signing

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Resolves `CryptoAlg-1536`

### Description of changes: 
Add supporting functions and tests for `OCSPRequest`

Taken from: 
* https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/ocsp/ocsp_cl.c

### Call-outs:
Code from OpenSSL was cleaned up to adhere to our code styling and library tools, but most of the actual logic stays the same. We disallow signing of the OCSP_REQUEST if the optional signature already exists.

### Testing:
* signing RSA certs and keys
* signing with ECDSA certs and keys
* signing if a signature already exists
* signing with a mismatched cert and key
* signing with a mismatched cert and key type

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

**Review by skmcgrail — DISMISSED** [review]

**Review by skmcgrail — DISMISSED** [review]

File: include/openssl/ocsp.h [hunk 1]
```diff
@@ -115,6 +119,22 @@ OPENSSL_EXPORT int OCSP_REQ_CTX_add1_header(OCSP_REQ_CTX *rctx,
 OPENSSL_EXPORT OCSP_ONEREQ *OCSP_request_add0_id(OCSP_REQUEST *req,
                                                  OCSP_CERTID *cid);
 
+// Set |requestorName| from an |X509_NAME| structure.
```
Comment (andrewhop) [inline include/openssl/ocsp.h:122]: Method name needs to go first in public documentation.

File: include/openssl/ocsp.h [hunk 2]
```diff
@@ -19,10 +19,14 @@
...
-#  define OCSP_NOCHAIN                    0x8
-#  define OCSP_NOEXPLICIT                 0x20
+// Various OCSP flags and values
+
+// Set OCSP_NOCERTS for OCSP_request_sign() if no certificates are included
+// in the |OCSP_REQUEST|. Certificates are optional.
+#define OCSP_NOCERTS                    0x1
+#define OCSP_NOINTERN                   0x2
+#define OCSP_NOCHAIN                    0x8
+#define OCSP_NOEXPLICIT                 0x20
```
Comment (andrewhop) [inline include/openssl/ocsp.h:29]: What is required for constant define documentation?

File: crypto/ocsp/ocsp_test.cc [hunk 3]
```diff
@@ -769,6 +897,53 @@ TEST_P(OCSPRequestTest, OCSPRequestParse) {
...
+  bssl::UniquePtr<X509> server_cert(CertFromPEM(
+      GetTestData(std::string("crypto/ocsp/test/aws/" +
+                        std::string(t.signer_cert) +".pem").c_str()).c_str()));
+  ASSERT_TRUE(server_cert);
+  bssl::UniquePtr<STACK_OF(X509)> additional_cert(CertChainFromPEM(
+      GetTestData(std::string("crypto/ocsp/test/aws/ca_cert.pem").c_str())
+          .c_str()));
+  ASSERT_TRUE(additional_cert);
+
+  if (t.expected_parse_status == OCSP_REQUEST_PARSE_SUCCESS) {
```
Comment (andrewhop) [inline crypto/ocsp/ocsp_test.cc:918]: Should this test the not success path?

File: crypto/ocsp/ocsp_client.c [hunk 4]
```diff
@@ -3,22 +3,113 @@
...
+  return 1;
+}
+
+int OCSP_request_sign(OCSP_REQUEST *req,
+                      X509 *signer,
+                      EVP_PKEY *key,
+                      const EVP_MD *dgst,
+                      STACK_OF(X509) *certs, unsigned long flags) {
+  if (req->optionalSignature != NULL) {
+    // OpenSSL lets an |OCSP_REQUEST| be signed twice, regardless of whether
```
Comment (andrewhop) [inline crypto/ocsp/ocsp_client.c:67]: 1. Just curious how did you find this edge case?
2. Do we have a test that attempts to sign twice?
3. Is this worthy of a new error message?

File: crypto/ocsp/ocsp_client.c [hunk 5]
```diff
@@ -3,22 +3,113 @@
 
 #include "internal.h"
 
-OCSP_ONEREQ *OCSP_request_add0_id(OCSP_REQUEST *req, OCSP_CERTID *cid)
-{
-    OCSP_ONEREQ *one = OCSP_ONEREQ_new();
-    if (one == NULL) {
-      return NULL;
+OCSP_ONEREQ *OCSP_request_add0_id(OCSP_REQUEST *req, OCSP_CERTID *cid) {
```
Comment (andrewhop) [inline crypto/ocsp/ocsp_client.c:6]: Do we have existing tests for these get/set/add functions?

File: include/openssl/ocsp.h (hunk 2 above)
Comment (samuel40791765) [reply include/openssl/ocsp.h]: The other three were probably missed out in previous imports. I can add documentation for them.

File: crypto/ocsp/ocsp_test.cc (hunk 3 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:918]: We're already testing this in `TEST_P(OCSPRequestTest, OCSPRequestParse)`(L730-L768) , so I opted not to make parse testing redundant.

File: crypto/ocsp/ocsp_client.c (hunk 4 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:67]: 1. I was trying to sign the signed OCSP requests test cases we have (`ocsp_request_signed`, `ocsp_request_signed_sha256`, etc.) with `OCSP_request_sign`, but I couldn't get past the ASAN build. Initially I couldn't make sense of it, but I dove a bit deeper and found that the original `req->optionalSignature` was being left behind.
2. If we try signing the signed OCSP requests test cases again, we'll be hitting this and fail. I've included those cases in `kRequestTestVectors`. 
3. I had been pondering that too, I'll add it.

File: crypto/ocsp/ocsp_client.c (hunk 5 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:6]: 1. For `OCSP_request_add0_id` we already have a test [here](https://github.com/aws/aws-lc/blob/ocsp-request/crypto/ocsp/ocsp_test.cc#L809-L840) in the branch.
2. As for `OCSP_request_set1_name` and `OCSP_request_add1_cert`, I assumed we already had coverage for  since `OCSP_request_sign` relies on them succeeding. I could add a more targeted test for them if needed?

**Review by andrewhop — APPROVED** [review]

**Review by skmcgrail — APPROVED** [review]
